File: php/src/Model/PrimitiveProperty.php
<?php

declare(strict_types=1);

namespace Oojs\Model;

/**
 * A property whose value is one of the JSON primitive types:
 * string, integer, number, boolean, null.
 */
class PrimitiveProperty implements PropertyDef
{
    /**
     * @param string          $kind             One of: string, integer, number, boolean, null
     * @param string          $title
     * @param string          $description
     * @param int|null        $minLength        String: minimum length (Unicode code points)
     * @param int|null        $maxLength        String: maximum length (Unicode code points)
     * @param string|null     $pattern          String: ECMA regex pattern
     * @param string|null     $format           String: informational format hint (not enforced v1.0)
     * @param float|null      $minimum          Numeric: inclusive minimum
     * @param float|null      $maximum          Numeric: inclusive maximum
     * @param float|null      $exclusiveMinimum Numeric: exclusive minimum
     * @param float|null      $exclusiveMaximum Numeric: exclusive maximum
     * @param float|null      $multipleOf       Numeric: divisor constraint (> 0)
     * @param list<mixed>|null $enum            Allowed values enumeration
     */
    public function __construct(
        public readonly string $kind,
        public readonly string $title = '',
        public readonly string $description = '',
        public readonly ?int $minLength = null,
        public readonly ?int $maxLength = null,
        public readonly ?string $pattern = null,
        public readonly ?string $format = null,
        public readonly ?float $minimum = null,
        public readonly ?float $maximum = null,
        public readonly ?float $exclusiveMinimum = null,
        public readonly ?float $exclusiveMaximum = null,
        public readonly ?float $multipleOf = null,
        public readonly ?array $enum = null,
    ) {}
}

File: java/src/main/java/org/oojs/model/PrimitiveProperty.java
package org.oojs.model;

import java.util.List;

/** A property whose value is one of the JSON primitive types: string, integer, number, boolean, null. */
public final class PrimitiveProperty implements PropertyDef {
    public final String kind;
    public final String title;
    public final String description;
    public final Integer minLength;
    public final Integer maxLength;
    public final String pattern;
    public final String format;
    public final Double minimum;
    public final Double maximum;
    public final Double exclusiveMinimum;
    public final Double exclusiveMaximum;
    public final Double multipleOf;
    public final List<Object> enumValues;

    public PrimitiveProperty(String kind, String title, String description,
            Integer minLength, Integer maxLength, String pattern, String format,
            Double minimum, Double maximum, Double exclusiveMinimum, Double exclusiveMaximum,
            Double multipleOf, List<Object> enumValues) {
        this.kind = kind;
        this.title = title;
        this.description = description;
        this.minLength = minLength;
        this.maxLength = maxLength;
        this.pattern = pattern;
        this.format = format;
        this.minimum = minimum;
        this.maximum = maximum;
        this.exclusiveMinimum = exclusiveMinimum;
        this.exclusiveMaximum = exclusiveMaximum;
        this.multipleOf = multipleOf;
        this.enumValues = enumValues;
    }
}
